FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean

```diff
@@ -196,28 +196,123 @@ noncomputable def adicCompletionComapSemialgHom (v : HeightOneSpectrum A) (w : H
 
+def comap_algebra_continuousSmul (v : HeightOneSpectrum A) (w : HeightOneSpectrum B)
+  (hvw : comap A w = v) : letI := comap_algebra A K L B hvw
+    ContinuousSMul (adicCompletion K v) (adicCompletion L w) := by
+  let inst_alg := comap_algebra A K L B hvw
+  constructor
+  have leftCts := adicCompletionComapSemialgHom_continuous A K L B v w hvw
+  exact Continuous.mul (Continuous.fst' leftCts) continuous_snd
+
+open TensorProduct in
+noncomputable def baseChange_of_algebraMap_adicComletion (v : HeightOneSpectrum A)
+    (w : HeightOneSpectrum B) (hvw : comap A w = v) :
+  letI := comap_algebra A K L B hvw
+  ((HeightOneSpectrum.adicCompletion K v) ⊗[K] L)
+      →ₗ[(HeightOneSpectrum.adicCompletion K v)] adicCompletion L w :=
+  let inst_alg := comap_algebra A K L B hvw
+  let sa : L →ₛₐ[algebraMap K (adicCompletion K v)] adicCompletion L w := {
+    __ := (algebraMap L (adicCompletion L w))
+    map_smul' x y := by
+      simp only [Algebra.smul_def, Algebra.smul_def, MonoidHom.map_mul']
+      congr 1
+      symm
+      apply (adicCompletionComapSemialgHom A K L B v w hvw).commutes x
+  }
+  (SemialgHom.baseChange_of_algebraMap sa).toLinearMap
+
+omit [IsIntegralClosure B A L] [Algebra.IsSeparable K L] [Module.Finite A B] in
+theorem baseChange_of_algebraMap_adicComletion_surjective (v : HeightOneSpectrum A)
+    (w : HeightOneSpectrum B) (hvw : comap A w = v) :
+    Function.Surjective (baseChange_of_algebraMap_adicComletion A K L B v w hvw) := by
+  let inst_alg := comap_algebra A K L B hvw
+  let inst_cSmul : ContinuousSMul (adicCompletion K v) (adicCompletion L w) :=
+    comap_algebra_continuousSmul A K L B v w hvw
+  set f := (baseChange_of_algebraMap_adicComletion A K L B v w hvw)
+  have isClosed : IsClosed (LinearMap.range f).carrier := by
+    apply Submodule.closed_of_finiteDimensional
+  let coeL : WithVal (w.valuation L) → adicCompletion L w := UniformSpace.Completion.coe'
+  have denseL : DenseRange coeL := UniformSpace.Completion.denseRange_coe
+  rw [← LinearMap.range_eq_top, Submodule.eq_top_iff']
+  simp_rw [← Submodule.mem_toAddSubmonoid, ← AddSubmonoid.mem_toSubsemigroup,
+      ← AddSubsemigroup.mem_carrier]
+  rw [← isClosed.closure_eq]
+  apply Dense.mono _ denseL
+  intro x ⟨l, hl⟩
+  use (1 ⊗ₜ l)
+  simp only [f, baseChange_of_algebraMap_adicComletion, ← hl,
+    SemialgHom.baseChange_of_algebraMap_tmul_right, AlgHom.toLinearMap_apply]
+  rfl
+
+omit [IsIntegralClosure B A L] [Algebra.IsSeparable K L] [Module.Finite A B] in
+theorem comap_algebra_finite (v : HeightOneSpectrum A) (w : HeightOneSpectrum B)
+    (hvw : comap A w = v)
+  : letI := comap_algebra A K L B hvw
+    Module.Finite (adicCompletion K v) (adicCompletion L w) := by
+  let inst_alg := comap_algebra A K L B hvw
+  have fdRange : (LinearMap.range (baseChange_of_algebraMap_adicComletion A K L B v w hvw)).FG := by
+    rw [← Module.Finite.iff_fg]
+    apply LinearMap.finiteDimensional_range
+  rw [Module.finite_def]
+  convert fdRange
+  symm
```

Comment:
```suggestion
  suffices LinearMap.range (baseChange_of_algebraMap_adicComletion A K L B v w hvw) = ⊤ by
    rwa [← this]
```
It's longer, but "`convert` + stuff" proofs scare me a bit; my suggestion will I think be easier to fix if the proof ever breaks after an update for some reason.